AdminServer is now the only one to send_webhook

farskipper opened the pull request:
First stab at the refactor. All the tests are passing and seems to be working fine with the toolbox.

Any event with a topic `acapy::webhook::(.*)` will get forwarded as a webhook. However, I'm thinking now might be a good time to improve the naming convention of these events, while still maintaining backwards compatibility with webhook topics.

i.e.  a mapping from event name to webhook topic might be:
```
{
  "acapy::basicmessage_received": "basicmessages",
  "acapy::ping": "ping",
  "acapy::ping_response": "ping",
  "acapy::driver_get_active_menu": "get-active-menu",
  "acapy::driver_perform_menu_action": "perform-menu-action",
  "acapy::problem_report": "problem_report",
}
```

farskipper committed:
AdminServer is now the only one to send_webhook

Signed-off-by: Matthew Wright <matthew@smallhelm.com>

farskipper commented on aries_cloudagent/protocols/actionmenu/v1_0/base_service.py:
```diff
@@ -28,19 +29,24 @@ async def get_instance(context: InjectionContext):
     async def get_active_menu(
-        self, connection: ConnRecord = None, thread_id: str = None
+        self,
+        profile: Profile,
```
Not sure if this is the best way to get access to the profile for the event.notify.

dbluhm replied:
The action menu is a bit of a strange case. I'm not sure just yet how we should handle it. Is a profile available easily in the contexts these methods are called from?

farskipper replied:
Yes, in ACA-Py they are only called in handlers

`get_active_menu` is called in MenuRequestHandler
`perform_menu_action` is called in PerformHandler

farskipper commented on aries_cloudagent/protocols/actionmenu/v1_0/base_service.py:
```diff
@@ -28,19 +29,24 @@ async def get_instance(context: InjectionContext):
     @abstractmethod
     async def perform_menu_action(
         self,
+        profile: Profile,
```
Not sure if this is the best way to get access to the profile for the event.notify.

farskipper commented on aries_cloudagent/core/event_bus.py:
```diff
@@ -45,7 +48,7 @@ def __init__(self):
         self.topic_patterns_to_subscribers: Dict[Pattern, Sequence[Callable]] = {}
 
-    async def notify(self, profile: Profile, event: Event):
+    async def notify(self, profile: "Profile", event: Event):
```
What do you think of having profile be an optional property of Event? Is there a specific reason that all events need a profile?

dbluhm replied:
Profile represents the wallet that emitted and should therefore process the event. This really only comes into play for multi-tenanted agents where one wallet manages many wallets. Omitting profile, the event handlers wouldn't have a handle to the correct storage and wallet to process the event.

farskipper replied:
Ok, makes sense that most events (in fact all of the ones we currently have) need a profile. However, do you think in the future we might need system wide events that are not specific to a wallet?

i.e.

`event_bus.notify(context.profile, Event("topic", payload))`
becomes
`event_bus.notify(Event("topic", payload, profile=context.profile))`

The annoying thing will be that most subscribers will need to assert that profile is present when handling the event.

dbluhm replied:
Fortunately, I think this is already kind of being taken care of given that there is the concept of a "root" profile on startup at least. Beyond that, I can't think of many scenarios where we would need system wide event since basically everything in ACA-Py is scoped to a wallet and storage, which must be provided by the profile. So I'm inclined to say that we skip it for now and, if it becomes obvious that is is needed in the future, we can worry about it then.

farskipper replied:
Ok, sounds good

dbluhm commented:
> However, I'm thinking now might be a good time to improve the naming convention of these events, while still maintaining backwards compatibility with webhook topics.
>
> i.e. a mapping from event name to webhook topic might be:
>
> ```
> {
>   "acapy::basicmessage_received": "basicmessages",
>   "acapy::ping": "ping",
>   "acapy::ping_response": "ping",
>   "acapy::driver_get_active_menu": "get-active-menu",
>   "acapy::driver_perform_menu_action": "perform-menu-action",
>   "acapy::problem_report": "problem_report",
> }
> ```

I agree with this approach with the one suggestion that adding more "namespaces" could be nice. So `acapy::basicmessage_received` might be `acapy::basicmessage::received`.

dbluhm commented on aries_cloudagent/admin/base_server.py:
```diff
@@ -23,20 +20,3 @@ async def start(self) -> None:
     @abstractmethod
     async def stop(self) -> None:
         """Stop the webserver."""
-
-    @abstractmethod
-    def add_webhook_target(
-        self,
-        target_url: str,
-        topic_filter: Sequence[str] = None,
-        max_attempts: int = None,
-    ):
-        """Add a webhook target."""
-
-    @abstractmethod
-    def remove_webhook_target(self, target_url: str):
-        """Remove a webhook target."""
-
-    @abstractmethod
-    async def send_webhook(self, profile: Profile, topic: str, payload: dict):
-        """Add a webhook to the queue, to send to all registered targets."""
```
These methods enable users to add webhook targets at run time through the Admin API. Since the webhook emitter is still a valid consumer of the event bus, I think it makes sense to leave this functionality as is. This will help us maintain backwards compatibility while still making webhooks happen as a result of an event instead of the other way around.

farskipper replied:
As I traced the usage, I noticed the webhook_targets were not used by the admin server. It looks like it's been superseded by `profile.settings.get("admin.webhook_urls")`

The code that uses the webhook targets added by those methods has been commented out:

https://github.com/Indicio-tech/aries-cloudagent-python/blame/10b8134ddb9bc6a26058c929c35ec7b230787174/aries_cloudagent/admin/server.py#L747-L748

dbluhm commented on aries_cloudagent/core/conductor.py:
```diff
@@ -138,10 +138,6 @@ async def setup(self):
                     self.dispatcher.task_queue,
                     self.get_stats,
                 )
-                webhook_urls = context.settings.get("admin.webhook_urls")
-                if webhook_urls:
-                    for url in webhook_urls:
-                        self.admin_server.add_webhook_target(url)
```
Setting webhook targets on startup should still be supported out of the box. See my previous comment. Webhooks are still first-class citizens.

dbluhm commented on aries_cloudagent/core/dispatcher.py:
```diff
@@ -298,16 +289,3 @@ async def send_outbound(self, message: OutboundMessage):
             message: The `OutboundMessage` to be sent
         """
         await self._send(self._context.profile, message, self._inbound_message)
-
-    async def send_webhook(self, topic: str, payload: dict):
-        """
-        Dispatch a webhook.
-
-        Args:
-            topic: the webhook topic identifier
-            payload: the webhook payload value
-        """
-        bus: EventBus = self._context.inject(EventBus)
-        await bus.notify(self._context.profile, Event(topic, payload))
-        if self._webhook:
-            await self._webhook(self._context.profile, topic, payload)
```
I think the responders should retain an `async def send_webhook` method that triggers a deprecation warning but still emits an event that will ultimately trigger a webhook.

farskipper replied:
@dbluhm makes sense. Should I use https://docs.python.org/3/library/warnings.html ?
i.e.
```py
warnings.warn("responder.send_webhook is deprecated; use event_bus instead.", DeprecationWarning)
```

dbluhm replied:
Yep, I think that's the way to go

dbluhm commented on aries_cloudagent/core/event_bus.py:
```diff
@@ -1,11 +1,14 @@
 """A simple event bus."""
 
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from ..core.profile import Profile
+
```
Oh interesting, I hadn't seen this pattern before. While it makes sense to box up imports that are used only for type hints, I think for now we ought to maintain ACA-Py convention and just let them be imported.

farskipper replied:
I added this this to avoid circular import error. AFAIK this is the way to get around it in python.

dbluhm replied:
Ah, I see. I missed that originally. Looks good to me then.

dbluhm commented on aries_cloudagent/core/profile.py:
```diff
@@ -99,6 +100,12 @@ async def close(self):
     async def remove(self):
         """Remove the profile."""
 
+    async def notify(self, topic: str, payload: Any):
+        """Signal an event."""
+        event_bus = self.inject(EventBus, required=False)
+        if event_bus:
+            await event_bus.notify(self, Event(topic, payload))
+
```
I hadn't thought to add this but I like it!

dbluhm commented on aries_cloudagent/messaging/responder.py:
```diff
@@ -113,24 +113,13 @@ async def send_outbound(self, message: OutboundMessage):
             message: The `OutboundMessage` to be sent
         """
 
-    @abstractmethod
-    async def send_webhook(self, topic: str, payload: dict):
-        """
-        Dispatch a webhook.
-
-        Args:
-            topic: the webhook topic identifier
-            payload: the webhook payload value
-        """
-
```
See my previous comment on responders retaining a send_webhook method.

dbluhm commented:
Looking pretty good so far. Left some comments and thoughts.

farskipper committed:
adding back responder.send_webhook with a deprecation warning

Signed-off-by: Matthew Wright <matthew@smallhelm.com>

farskipper committed:
mapping events to webhooks

Signed-off-by: Matthew Wright <matthew@smallhelm.com>

farskipper commented:
Made some changes:

- Brought back `responder.send_webhook` with a deprecation warning
- Added comment to explain why `if TYPE_CHECKING` is used
- Added `EVENT_WEBHOOK_MAPPING`

Question: Is it ok to remove these?
- `BaseAdminServer.add_webhook_target` - superseded by `profile.settings.get("admin.webhook_urls")`
- `BaseAdminServer.remove_webhook_target` - superseded by `profile.settings.get("admin.webhook_urls")`
- `BaseAdminServer.send_webhook` - Now that we have events, this seems like an implementation detail, not necessarily something all admin servers need to support.

dbluhm commented:
> BaseAdminServer.send_webhook - Now that we have events, this seems like an implementation detail, not necessarily something all admin servers need to support.

Sounds good to me

> - BaseAdminServer.add_webhook_target - superseded by profile.settings.get("admin.webhook_urls")
> - BaseAdminServer.remove_webhook_target - superseded by profile.settings.get("admin.webhook_urls")

I see, I didn't pick up on the `self.webhook_urls` not even being used in the `send_webhook` method; I defer to your findings.

dbluhm commented:
Apologies, I've lost track of all the different conversations -- are there further changes needed after our discussion or is this ready for final review?

farskipper commented:
@dbluhm I think it's ready for final review if we are ok removing `BaseAdminServer.[add/remove]_webhook_target` All the tests are passing and seems to work fine with the toolbox.

Are there other implementations of BaseAdminServer running around out there? If so we'll just have to note that it's a breaking change to rely on `add_webhook_target` and you should be using `profile.settings.get("admin.webhook_urls")` instead. Which I think is fine b/c that's all it's really using right now anyways.

dbluhm reviewed (changes requested):
Nice work! Looking really good!

Looking through, I think we should generalize the `send_webhook` method on `BaseRecord`. At the moment, record updates (which as it turns out is one of the most useful and frequent types of webhook event sent out) are emitting an event specifically scoped to the webhook.

I recommend:

1. Rename `BaseRecord.send_webhook` to something event oriented, perhaps `emit_event` or similar.
2. Use `acapy::RECORD_TOPIC::STATE` as the topic.
   - This will require adding mappings as we have for the basic messages and pings.

This one might get a little messy but I think I would also recommend modifying the `BaseRecord.save` and `BaseRecord.post_save` (and any other relevant) methods to take a bool `event` instead of `webhook` or else keep webhook around for backwards compatibility and add the event but make sure they are the same and emit a deprecation warning on webhook usage. I'm not bent on this one but I think abstracting away the webhook as much as possible will make the system cleaner overall. If this ends up being a huge and/or brittle change, we'll likely need to regroup and come up with a different solution.

farskipper committed:
base record replace webhooks with events

Signed-off-by: Matthew Wright <matthew@smallhelm.com>

farskipper commented:
@dbluhm Ok, I made a PR to make those changes: #59 I did a PR so it'd be easy to diff and back out if we decide against it.

farskipper committed:
split up admin server event subscribers

Signed-off-by: Matthew Wright <matthew@smallhelm.com>

dbluhm committed:
Merge pull request #59 from Indicio-tech/event-bus-webhook-refactor-base_record

Base record replace webhooks with events
